client/src/templates/Introduction/components/super-block-accordion.tsx

```diff
@@ -163,30 +200,61 @@ export const SuperBlockAccordion = ({
           ({ block: blockName }) => blockName === block.dashedName
         );
 
+        const completedBlockChallenges = completedChallengesIds.filter(
+          completedChallengeId =>
+            blockChallenges.some(
+              ({ id: blockChallengeId }) =>
+                blockChallengeId === completedChallengeId
+            )
+        );
+
         return {
           name: block.dashedName,
           blockType: blockChallenges[0]?.blockType ?? null,
-          challenges: blockChallenges
+          challenges: blockChallenges,
+          isCompleted:
+            completedBlockChallenges.length === blockChallenges.length
         };
       });
 
-    const allChapters = chapters.map(chapter => ({
-      name: chapter.dashedName,
-      comingSoon: chapter.comingSoon,
-      modules: chapter.modules.map((module: Module) => ({
-        name: module.dashedName,
-        comingSoon: module.comingSoon,
-        blocks: populateBlocks(module.blocks)
-      }))
-    }));
+    const populateModules = (modules: Module[]) =>
+      modules.map((module: Module) => {
+        const blocks = populateBlocks(module.blocks);
+        const isCompleted = blocks.every(block => block.isCompleted === true);
+
+        return {
+          name: module.dashedName,
+          comingSoon: module.comingSoon,
+          blocks,
+          isCompleted
+        };
+      });
+
+    const allChapters = chapters.map(chapter => {
+      const modules = populateModules(chapter.modules);
+      const isCompleted = modules.every(module => module.isCompleted === true);
+
+      return {
+        name: chapter.dashedName,
+        comingSoon: chapter.comingSoon,
+        modules: populateModules(chapter.modules),
+        isCompleted
+      };
+    });
 
     return { allChapters };
-  }, [challenges]);
+  }, [challenges, completedChallengesIds]);
 
   // Expand the outer layers in order to reveal the chosen block.
   const expandedChapter = blockToChapterMap.get(chosenBlock);
   const expandedModule = blockToModuleMap.get(chosenBlock);
 
+  // Save allChapters to redux
+  useEffect(() => {
+    updateAllChapters(allChapters);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```

Comment:
I think we need to look at this from the perspective of completed modules and chapters (and perhaps blocks) in isolation rather than starting from the challenges.

It comes back to what Tom said in the meeting yesterday: we need completion state to survive the addition of new challenges. If we use completed challenges to derive `block.isCompleted`, then if we push a new challenge that would flip that and `isCompleted` for its parent module and again for its parent chapter.

What makes sense to me is ask the api for the completion state and use that to populate the redux store. The client then asks the api to update completion state (at the end of a section) and adjusts the store based on what the api returns.  In other words, the api is the single source of truth for completion state and the client just renders what it renders based on that state.

Does that seem reasonable?

Reply:
Ah, I'm with you. 

Importing `full-stack.json` should be fine - it's the source of truth. Then you can derive the completion state from the completed challenges. It's kind of an anti-pattern to derive state, but it makes sense here since we can just rip out the derivation logic and insert the completion state once that's returned by the api.